shared/p2p/service.go

```diff
@@ -91,6 +77,85 @@ func (s *Server) Stop() error {
 	return nil
 }
 
+// RegisterTopic with a message and the adapter stack for the given topic. The
+// message type provided will be feed selector for emitting messages received
+// on a given topic.
+//
+// The topics can originate from multiple sources. In other words, messages on
+// TopicA may come from direct peer communication or a pub/sub channel.
+func (s *Server) RegisterTopic(topic string, message interface{}, adapters ...Adapter) {
```

Comment:
I tried this and it's failing tests. I wonder if reflect is assuming that TypeOf(message) is proto.Message rather than the full proto definition. 

I didn't investigate thoroughly.